### ISSUE TITLE
Support kms:ViaService for KMS key operations

### ISSUE THREAD

**Issue by iann0036** [aws-cloudformation/aws-cloudformation-coverage-roadmap#321]
It would be great if CloudFormation became a supported service for the kms:ViaService condition as per:

https://docs.aws.amazon.com/kms/latest/developerguide/policy-conditions.html#conditions-kms-via-service

This would allow me to create a KMS key policy that enforces that only operations performed by CloudFormation are permitted for certain actions (for example, encrypting Lambda function environment variables).

Categories: Security (KMS), Enhancement

**Comment by acdebaca** [comment]
Hi @iann0036:
This already works today. We have used this feature specifically for encrypting lambda function environment variables. For example:
```
       - Sid: CFRoleEncryptLambda
          Effect: Allow
          Principal:
            AWS: arn:aws:iam::123456789012:role/cloudformation-role
          Resource: "*" 
          Action:
          - kms:Encrypt
          Condition:
            StringEquals:
              kms:ViaService: cloudformation.us-east-1.amazonaws.com
```
...and you would also include a statement for the role assumed by the lambda function allowing `kms:Decrypt`, which could use `kms:ViaService: lambda.us-east-1.amazonaws.com`.

Perhaps you omitted the region in the DNS name in your own testing? It's not the same DNS name that you would include in an assume-role trust policy (e.g. for the lambda function), which does not include the region.

The one problem that we have encountered is that, if you are using [a dedicated CloudFormation service role](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-servicerole.html), then that role must have `kms:PutKeyPolicy` and `kms:CreateKey` specified on the key policy **without any conditions**, or you will receive an error like **The new key policy will not allow you to update the key policy in the future** -- even if you have granted these permissions to another role (e.g. a super user role). It's pretty arbitrary, in fact, which actions succeed and which fail when you specify conditions like `kms:ViaService`. YMMV, Just my $0.02.

**Comment by iann0036** [comment]
Thanks @acdebaca soo much!

This was indeed using a service role while making a [minimal privilege project](https://github.com/iann0036/cfn-leastprivilege) - I'll incorporate your feedback in.

I've submitted feedback on the documentation to AWS so hopefully they update the page.

**Comment by benkehoe** [comment]
Is this then a feature request to the KMS team for better key policies?

**Comment by iann0036** [comment]
I think they key policies are fine and it's doing the right thing by checking for key lockout.

This is now more of a KMS documentation issue, which I've raised feedback for - unless you think otherwise?

**Comment by acdebaca** [comment]
@iann0036
I know the issue is closed, but I discovered that I was mistaken in my previous post (which I've now edited). Because it is actually the lambda service that is performing the environment variable encryption and not the cloudformation service (which is only triggering the encryption), you'll want to use `lambda.{region}.amazonaws.com` with the `kms:ViaService` key.

**Comment by iann0036** [comment]
Ah, good point. I wonder if CloudFormation itself ever needs to do KMS operations directly and not on behalf of a service?

**Comment by acdebaca** [comment]
@iann0036 It's a good question. Bottom line is I don't know. I tried a few things but still cannot come up with a case in which it's needed.

I tested uploading a lambda to an [S3 bucket with SSE-KMS](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-serversideencryptionbydefault.html), all created via CloudFormation. The CloudFormation service role does appear to require `kms:Decrypt` on the bucket key in order to read the lambda source code, however in that case the `kms:ViaService` value that worked for me was `s3.region.amazonaws.com`. (Note that uploading to the bucket is not done by the CloudFormation service role).

In an attempt to cover a case where the CloudFormation service role needs to encrypt data with the key, I also tested creating [a generated secret](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-secretsmanager-secret.html#cfn-secretsmanager-secret-generatesecretstring) with a KMS CMK, again all created via CloudFormation. In this case, the CloudFormation service role required `kms:Decrypt` and `kms:GenerateDataKey` - [it does not use `kms:Encrypt`](https://docs.aws.amazon.com/kms/latest/developerguide/services-secrets-manager.html). In this case, the value for `kms:ViaService` that worked for me was `secretsmanager.region.amazonaws.com`.
